src/H5Odtype.c

```diff
@@ -388,6 +389,13 @@ H5O__dtype_decode_helper(unsigned *ioflags /*in,out*/, const uint8_t **pp, H5T_t
                 if (H5_IS_KNOWN_BUFFER_OVERFLOW(skip, *pp, actual_name_length, p_end))
                     HGOTO_ERROR(H5E_OHDR, H5E_OVERFLOW, FAIL, "ran off end of input buffer while decoding");
 
+                /* Check for duplicated field name */
+                for (memb_idx = 0; memb_idx < dt->shared->u.compnd.nmembs; memb_idx++)
```

Comment:
I wonder if we can do this check at the end of decoding a compound datatype rather than doing it for every single member. That seems like a real performance killer for large compounds.

Reply:
I believe that it'll be O(n^2) either way...  :-/

Reply:
What if we build a temporary hashmap indexed by hashing the member name and use it for lookups?

Reply:
Not something that really needs to hold this PR up, but there are probably some ways we can improve on this.